File: python-scripts/advertiseservice/advertiseservice.py
import datetime
import json
import logging as log
import time

import const


class AdvertiserService:
    def __init__(self):
        log.info('Setting up advertising service')
        self.global_ids = self._read_global_ids()
        self.last_keepalive = dict()
        self.connected = dict()

        self.autodeploy_data = self.read_autodeploy()
        if const.DEPLOY_SELF in self.autodeploy_data:
            self.host = self.autodeploy_data[const.DEPLOY_SELF]
        else:
            self.host = None
        log.info('Autodeploy data: ' + str(self.autodeploy_data))

        # calculate smallest timeout
        timeouts = [device[const.ADAPTER_CONF][const.TIMEOUT] for device in self.autodeploy_data[const.DEPLOY_DEVICES]]
        self.min_timeout = datetime.timedelta(seconds=min(timeouts)) / 2

        self.set_unconnected()

    def start(self, AdvertiserClass):
        log.info('Starting advertising service with |%s|', str(AdvertiserClass))

        advertiser = AdvertiserClass(self)

        tries = 0
        connected = False
        while not connected and tries < 5:
            tries += 1
            log.debug('advertising try |%d|', tries)
            advertiser.advertise()
            connected = self.check_connected()

        connected_devices = [(key, self.global_ids) for key in self.global_ids if self.global_ids[key]]
        if len(connected_devices) == 0:
            log.info('No devices connected exiting')
        else:
            log.info('Successfully connected |%d| devices', len(connected_devices))

        while True:
            cur_time = datetime.datetime.utcnow()
            for device_name in self.last_keepalive.keys():
                if not self.global_ids[device_name]:
                    log.debug('Device |%s| not connected', device_name)
                    continue
                last_contact = self.last_keepalive[device_name][const.LAST_CONTACT]
                timeout = self.last_keepalive[device_name][const.TIMEOUT]
                passed_time = cur_time - last_contact
                max_passed_time = timeout - self.min_timeout
                log.debug('Checking device |%s|. '
                          '\nTimeout is |%s|, min timeout is |%s|, max passed time is |%s|,'
                          '\nlast contact at |%s|, now is |%s|, passed time is |%s|,'
                          '\nneed keep alive |%s|',
                          device_name, str(timeout), str(self.min_timeout), str(max_passed_time), str(last_contact),
                          str(cur_time), str(passed_time), str(passed_time >= max_passed_time))
                if passed_time >= max_passed_time:
                    advertiser.send_keep_alive(device_name)
                    self.last_keepalive[device_name][const.LAST_CONTACT] = cur_time

            time.sleep(self.min_timeout.total_seconds())

    def stop(self):
        with open(const.GLOBAL_ID_FILE, 'w') as f:
            json.dump(self.global_ids, f)
        log.info('Stopping advertising service')

    def read_autodeploy(self):
        try:
            json_string = open(const.AUTODEPLOY_FILE).read()
            deploy_data = json.loads(json_string)
            return deploy_data
        except IOError:
            return None
        except json.JSONDecodeError:
            log.exception('Illformated autodeploy file')

    def set_unconnected(self):
        # TODO client - replace const.NAME with const.LOCAL_ID
        if self.host is not None:
            self.connected[self.host[const.NAME]] = False
            self.last_keepalive[self.host[const.NAME]] = {
                const.TIMEOUT: datetime.timedelta(seconds=self.host[const.ADAPTER_CONF][const.TIMEOUT]),
                const.LAST_CONTACT: datetime.datetime.utcnow()
            }
            if self.host[const.NAME] not in self.global_ids:
                self.global_ids[self.host[const.NAME]] = 0

        for device in self.autodeploy_data[const.DEPLOY_DEVICES]:  # set invalid id for all devices
            self.connected[device[const.NAME]] = False
            self.last_keepalive[device[const.NAME]] = {
                const.TIMEOUT: datetime.timedelta(seconds=device[const.ADAPTER_CONF][const.TIMEOUT]),
                const.LAST_CONTACT: datetime.datetime.utcnow()
            }
            if device[const.NAME] not in self.global_ids:
                self.global_ids[device[const.NAME]] = 0

    def check_connected(self):
        # TODO client - must be able to tell the difference between (un-)successfull connect and reconnect (reconnect already has a valid global_id)
        connected = True
        if self.host is not None and not self.global_ids[self.host[const.NAME]]:
            connected = False

        for device in self.autodeploy_data[const.DEPLOY_DEVICES]:
            if not self.global_ids.get(device[const.NAME]):
                connected = False

        return connected

    def _read_global_ids(self):
        try:
            with open(const.GLOBAL_ID_FILE, 'r') as file:
                return json.load(file)
        except IOError:
            log.exception('Error opening global_ids file')
        except json.JSONDecodeError:
            log.exception('Illformated autodeploy file')

        return dict()

File: ConndeApp/app/src/main/java/org/citopt/connde/conndeapp/advertise/AdvertiseService.java
package org.citopt.connde.conndeapp.advertise;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.Charset;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Created by rosso on 19.08.17.
 */

public class AdvertiseService {
  private static final Logger log = LoggerFactory.getLogger(AdvertiseService.class);

  private Map<String, AdvertiseDevice> devices; // TODO include host in devices?
  private DiscoveredServer discoveredServer;
  private List<ServerDiscoveredListener> discoveredListeners = new LinkedList<>();

  private JSONObject autodeploy_data;
  private AdvertiseDevice host;
  private double min_timeout; // timeout in seconds
  private AtomicBoolean keepalive = new AtomicBoolean(true);

  private File filesDir;

  public AdvertiseService(File filesDir) throws JSONException {
    log.info("Setting up advertising service");

    this.filesDir = filesDir;
    devices = new HashMap<>();

    read_autodeploy();
    read_global_ids();

    log.info("Autodeploy data: " + this.autodeploy_data.toString());

    // calculate smallest timeout
    JSONArray devices = autodeploy_data.getJSONArray(Const.DEPLOY_DEVICES);
    min_timeout = Integer.MAX_VALUE;
    for (int i = 0; i < devices.length(); i++) {
      JSONObject device = devices.getJSONObject(i);
      int timeout = device.getJSONObject(Const.ADAPTER_CONF).getInt(Const.TIMEOUT);
      min_timeout = Math.min(min_timeout, timeout);
    }
    min_timeout = (min_timeout / 2) * 1000;
  }

  private void read_global_ids() throws JSONException {
    File globalIdFile = new File(filesDir, Const.GLOBAL_ID_FILE);
    JSONObject readObject = null;
    if (globalIdFile.exists()) {
      log.info("Found GLOBAL_ID file. Reading for reconnection...");
      try (InputStream is = new FileInputStream(globalIdFile)) {
        int size = is.available();
        byte[] buffer = new byte[size];
        is.read(buffer);
        String json = new String(buffer, "UTF-8");
        readObject = new JSONObject(json);
      } catch (FileNotFoundException e) {
        e.printStackTrace();
      } catch (IOException e) {
        e.printStackTrace();
      }
    } else {
      log.info("Could not find GLOBAL_ID file.");
      readObject = new JSONObject();
    }

    Iterator<String> keys = readObject.keys();
    while (keys.hasNext()) {
      String localId = keys.next();
      int globalId = readObject.getInt(localId);

      setGlobalId(localId, globalId);
    }
  }

  private void read_autodeploy() throws JSONException {
    File autodeployFile = new File(filesDir, Const.AUTODEPLOY_FILE);
    JSONObject readObject = null;
    if (autodeployFile.exists()) {
      try (InputStream is = new FileInputStream(autodeployFile)) {
        int size = is.available();
        byte[] buffer = new byte[size];
        is.read(buffer);
        is.close();
        String json = new String(buffer, "UTF-8");
        readObject = new JSONObject(json);
      } catch (IOException ex) {
        ex.printStackTrace();
      }
    } else {
      readObject = new JSONObject();
    }

    autodeploy_data = readObject;


    if (autodeploy_data.has(Const.DEPLOY_SELF)) {
      this.host = DeviceHelper.deviceFromJSON(this.autodeploy_data.getJSONObject(Const.DEPLOY_SELF));
    } else {
      this.host = null;
    }

    if (autodeploy_data.has(Const.DEPLOY_DEVICES)) {
      JSONArray deviceArray = autodeploy_data.getJSONArray(Const.DEPLOY_DEVICES);
      for (int i = 0; i < deviceArray.length(); i++) {
        JSONObject jsonDevice = deviceArray.getJSONObject(i);
        AdvertiseDevice device = DeviceHelper.deviceFromJSON(jsonDevice);
        this.devices.put(device.getLocalId(), device);
      }
    }
  }

  private boolean check_connected() {
    boolean connected = true;

    connected = discoveredServer != null;

    if (host != null && !host.isConnected()) {
      connected = false;
    }

    for (String key : devices.keySet()) {
      AdvertiseDevice device = devices.get(key);
      if (!device.isConnected()) {
        connected = false;
      }
    }

    return connected;
  }

  public void start(String AdvertiserClass) throws JSONException {
    log.info("Starting advertising service with |{}|", AdvertiserClass);

    AdvertiseClient advertiser = new AndroidAdvertiser(Const.LAN, this);

    int tries = 0;
    boolean connected = false;
    while (!connected && tries < 5) {
      tries += 1;
      log.debug("advertising try |{}|", tries);
      DiscoveredServer newDiscoveredServer = advertiser.advertise();
      if (newDiscoveredServer != null) {
        setDiscoveredServer(newDiscoveredServer);
      }
      connected = this.check_connected();
    }

    int connected_devices = 0;
    for (String curKey : this.devices.keySet()) {
      if (this.devices.get(curKey).isConnected()) {
        connected_devices++;
      }
    }

    if (connected_devices == 0) {
      log.info("No devices connected exiting");
      return;
    } else {
      log.info("Successfully connected |{}| devices", connected_devices);
    }

    while (keepalive.get()) {
      long cur_time = System.currentTimeMillis();
      Map<String, AdvertiseDevice> allDevices = this.getAllDevices();
      for (String device_name : allDevices.keySet()) {
        AdvertiseDevice device = allDevices.get(device_name);
        if (!device.isConnected()) {
          log.debug("Device |{}| not connected", device_name);
          continue;
        }
        long last_contact = device.getLastKeepalive();
        double timeout = device.getAutodeployDouble(Const.TIMEOUT);
        long passed_time = cur_time - last_contact;
        double max_passed_time = timeout - this.min_timeout;
        log.debug("Checking device |{}|. " +
                        "\nTimeout is |{}|, min timeout is |{}|, max passed time is |{}|," +
                        "\nlast contact at |{}|, now is |{}|, passed time is |{}|," +
                        "\nneed keep alive |{}|",
                device_name, timeout, this.min_timeout, max_passed_time, last_contact,
                cur_time, passed_time, String.valueOf(passed_time >= max_passed_time));
        if (passed_time >= max_passed_time) {
          advertiser.send_keepalive(device_name);
          device.setLastKeepalive(cur_time);
        }
      }
      try {
        Thread.sleep((long) this.min_timeout);
      } catch (InterruptedException e) {
        e.printStackTrace();
      }
    }
  }

  public void stop() throws JSONException {
    keepalive.set(false);
    File globalIdFile = new File(filesDir, Const.GLOBAL_ID_FILE);
    try (OutputStream os = new FileOutputStream(globalIdFile)) {
      JSONObject writeObject = new JSONObject();
      Map<String, AdvertiseDevice> allDevices = getAllDevices();
      for (String localId : allDevices.keySet()) {
        int globalId = allDevices.get(localId).getGlobalId();
        if (globalId > 0) {
          writeObject.put(localId, globalId);
        }
      }
      os.write(writeObject.toString(2).getBytes(Charset.forName("UTF-8")));
      os.flush();
    } catch (FileNotFoundException e) {
      e.printStackTrace();
    } catch (IOException e) {
      e.printStackTrace();
    }
  }

  void setGlobalId(String deviceName, int globalId) {
    AdvertiseDevice device = devices.get(deviceName);
    if (device != null) {
      device.setGlobalId(globalId);
    } else if (host != null && host.getLocalId().equals(deviceName)) {
      host.setGlobalId(globalId);
    }
  }

  int getGlobalId(String deviceName) {
    AdvertiseDevice device = devices.get(deviceName);
    if (device != null) {
      return device.getGlobalId();
    } else if (host != null && host.getLocalId().equals(deviceName)) {
      return host.getGlobalId();
    }
    return -1;
  }

  void setConnected(String deviceName, boolean connected) {
    AdvertiseDevice device = devices.get(deviceName);
    if (device != null) {
      device.setConnected(connected);
    } else if (host != null && host.getLocalId().equals(deviceName)) {
      host.setConnected(connected);
    }
  }

  public AdvertiseDevice getHost() {
    return host;
  }

  public Map<String, AdvertiseDevice> getDevices() {
    return devices;
  }

  Map<String, AdvertiseDevice> getAllDevices() {
    if (host != null) {
      Map<String, AdvertiseDevice> allDevices = new HashMap<>(devices);
      allDevices.put(host.getLocalId(), host);
      return allDevices;
    } else {
      return devices;
    }

  }

  private void setDiscoveredServer(DiscoveredServer newDiscoveredServer) {
    discoveredServer = newDiscoveredServer;
    for (ServerDiscoveredListener curListener : discoveredListeners) {
      curListener.onServerDiscovered(new ServerDiscoveredListener.ServerDiscoveredEvent(discoveredServer));
    }
  }

  public DiscoveredServer getDiscoveredServer() {
    return discoveredServer;
  }

  public void addDiscoveredListener(ServerDiscoveredListener listener) {
    discoveredListeners.add(listener);
  }

  public void removeDiscoveredListener(ServerDiscoveredListener listener) {
    discoveredListeners.remove(listener);
  }
}
